File: calculate_distances.py
import numpy as np

def calculate_distances(r1, r2, dimensionality):
    r1_array = np.array(r1[:dimensionality])
    r2_array = np.array(r2[:dimensionality])
    return np.linalg.norm(r1_array - r2_array)

def calculate_all_distances(atom_coordinates, dimensionality):
    distance_map = {}
    number_of_points = len(atom_coordinates)
    for i in range(number_of_points):
        for j in range(i + 1, number_of_points):
            dist = calculate_distances(atom_coordinates[i], atom_coordinates[j], dimensionality)
            distance_map[(i, j)] = dist
            distance_map[(j, i)] = dist
    return distance_map

File: C++/calculate_distances.cpp
#include <cmath>  // For std::sqrt and std::pow
#include <vector>

using namespace std;

double calculate_distances(const vector<float>& r1, const vector<float>& r2, int dimensionality) {
    double sum = 0.0;
    for (int i = 0; i < dimensionality; ++i) {
        sum += pow(r1[i] - r2[i], 2);
    }
    return sqrt(sum);
}

#include <map>
#include <utility>  // For std::pair

map<pair<int, int>, double> calculate_all_distances(const vector<vector<float>>& atom_coordinates, int dimensionality) {
    map<pair<int, int>, double> distance_map;
    int number_of_points = atom_coordinates.size();
    for (int i = 0; i < number_of_points; ++i) {
        for (int j = i + 1; j < number_of_points; ++j) {
            double dist = calculate_distances(atom_coordinates[i], atom_coordinates[j], dimensionality);
            distance_map[make_pair(i, j)] = dist;
            distance_map[make_pair(j, i)] = dist;
        }
    }
    return distance_map;
}
